bitcoin/src/blockdata/transaction.rs

```diff
@@ -1148,6 +1161,107 @@ impl From<&Transaction> for Wtxid {
     }
 }
 
+/// Predicts the weight of a to-be-constructed transaction.
+///
+/// This function computes the weight of a transaction which is not fully known. All that is needed
+/// is the lengths of scripts and witness elements.
+///
+/// # Arguments
+///
+/// * `inputs` - an iterator which returns `InputWeightPrediction` for each input of the
+///   to-be-constructed transaction.
+/// * `output_script_lens` - an iterator which returns the length of `script_pubkey` of each input
+///   of the to-be-constructed transaction.
+///
+/// # Usage
+///
+/// When signing a transaction one doesn't know the signature before knowing the transaction fee and
+/// the transaction fee is not known before knowing the transaction size which is not known before
+/// knowing the signature. This apparent dependency cycle can be broken by knowing the length of the
+/// signature without knowing the contents of the signature e.g., we know all Schnorr signatures
+/// are 64 bytes long.
+///
+/// Additionally, some protocols may require calculating the amounts before knowing various parts
+/// of the transaction (assuming their length is known).
+///
+/// # Notes on integer overflow
+///
+/// Overflows are intentionally not checked because one of the following holds:
+///
+/// * The transaction is valid (obeys the block size limit) and the code feeds correct values to
+///   this function - no overflow can happen.
+/// * The transaction will be so large it doesn't fit in the memory - overflow will happen but
+///   then the transaction will fail to construct and even if one serialized it on disk directly
+///   it'd be invalid anyway so overflow doesn't matter.
+/// * The values fed into this function are inconsistent with the actual lengths the transaction
+///   will have - the code is already broken and checking overflows doesn't help. Unfortunately
+///   this probably cannot be avoided.
+pub fn predict_weight<I, O>(inputs: I, output_script_lens: O) -> Weight
```

Comment:
I also thought about this a bit. I think in a later PR we can maybe make some constructors for these iterators which can be created off of a (potentially incomplete) transaction.

Agree that in the majority of cases users will be starting with some sort of transaction and trying to estimate its final form.

Reply:
I had the same idea and was about to make the method for output iterator but realized that I actually don't have this flow because it would require to add an output with bogus amount (presumably 0) and then change it... Which is not really that nice. I prefer to calculate everything first and only then construct the transaction. But I can add the method back if you want, it's one line of code.

Reply:
I think it'd be worthwhile, yeah. I think rust-miniscript is going to use this method, and in Miniscript we assume that all inputs and outputs are present, just (potentially) not witnesses.

Reply:
OK, added.